op-program/client/tasks/deposits_block.go

```diff
@@ -0,0 +1,151 @@
+)
+
+// BuildDepositOnlyBlock builds a deposits-only block replacement for the specified optimistic block and returns the block hash and output root
+// for the new block.
+// The specified l2OutputRoot must be the output root of the optimistic block's parent.
+func BuildDepositOnlyBlock(
+	logger log.Logger,
+	cfg *rollup.Config,
+	l2Cfg *params.ChainConfig,
+	optimisticBlock *types.Block,
+	l1Head common.Hash,
+	agreedL2OutputRoot eth.Bytes32,
+	l1Oracle l1.Oracle,
+	l2Oracle l2.Oracle,
+) (common.Hash, eth.Bytes32, error) {
+	engineBackend, err := l2.NewOracleBackedL2Chain(logger, l2Oracle, l1Oracle, l2Cfg, common.Hash(agreedL2OutputRoot))
+	if err != nil {
+		return common.Hash{}, eth.Bytes32{}, fmt.Errorf("failed to create oracle-backed L2 chain: %w", err)
+	}
+	l2Source := l2.NewOracleEngine(cfg, logger, engineBackend)
+	l2Head := l2Oracle.BlockByHash(optimisticBlock.ParentHash(), eth.ChainIDFromBig(l2Cfg.ChainID))
+	l2HeadHash := l2Head.Hash()
+
+	optimisticBlockOutput, err := getL2Output(logger, cfg, l2Cfg, l2Oracle, l1Oracle, optimisticBlock)
+	if err != nil {
+		return common.Hash{}, eth.Bytes32{}, fmt.Errorf("failed to get L2 output: %w", err)
+	}
+	logger.Info("Building a deposts-only block to replace block %v", optimisticBlock.Hash())
+	attrs, err := blockToDepositsOnlyAttributes(cfg, optimisticBlock, optimisticBlockOutput)
+	if err != nil {
+		return common.Hash{}, eth.Bytes32{}, fmt.Errorf("failed to convert block to deposits-only attributes: %w", err)
+	}
+	result, err := l2Source.ForkchoiceUpdate(context.Background(), &eth.ForkchoiceState{
+		HeadBlockHash:      l2HeadHash,
+		SafeBlockHash:      l2HeadHash,
+		FinalizedBlockHash: l2HeadHash,
+	}, attrs)
+	if err != nil {
+		return common.Hash{}, eth.Bytes32{}, fmt.Errorf("failed to update forkchoice state: %w", err)
+	}
+	if result.PayloadStatus.Status != eth.ExecutionValid {
+		return common.Hash{}, eth.Bytes32{}, fmt.Errorf("failed to update forkchoice state: %w", eth.ForkchoiceUpdateErr(result.PayloadStatus))
+	}
+
+	id := result.PayloadID
+	if id == nil {
+		return common.Hash{}, eth.Bytes32{}, fmt.Errorf("nil id in forkchoice result when expecting a valid ID")
+	}
+	payload, err := l2Source.GetPayload(context.Background(), eth.PayloadInfo{ID: *id, Timestamp: uint64(attrs.Timestamp)})
+	if err != nil {
+		return common.Hash{}, eth.Bytes32{}, fmt.Errorf("failed to get payload: %w", err)
+	}
+	blockHash, outputRoot, err := l2Source.L2OutputRoot(uint64(payload.ExecutionPayload.BlockNumber))
+	if err != nil {
+		return common.Hash{}, eth.Bytes32{}, fmt.Errorf("failed to get L2 output root: %w", err)
+	}
+	return blockHash, outputRoot, nil
+}
+
+func getL2Output(logger log.Logger, cfg *rollup.Config, l2Cfg *params.ChainConfig, l2Oracle l2.Oracle, l1Oracle l1.Oracle, block *types.Block) (*eth.OutputV0, error) {
+	backend := l2.NewOracleBackedL2ChainFromHead(logger, l2Oracle, l1Oracle, l2Cfg, block)
+	engine := l2.NewOracleEngine(cfg, logger, backend)
```

Comment:
Shouldn't we be able to reuse these from the ones already created in `BuildDepositOnlyBlock`?

Reply:
The deposits only block chain doesn't have a full view of the state needed to construct the optimistic output. So we need to rebuild a new engine, rooted at the optimistic block hash, that has all that data.

Reply:
We roll back our chain view to the agreed state for the block replacement. If the block replacement contains any withdrawals, then we won't be able to reconstruct the optimistic output root.

Reply:
Ah right, the optimistic block hasn't been imported into the engine at all. Logically it does then make sense to create a new version.  I'd be interested to know if it would work without it though - I have a sneaking suspicion that because the required preimages are available via the oracle that it would just work (geth determines if a world state is available or not just by trying to load the root).

Reply:
But just to be clear, I agree with keeping the code as-is with the extra creation. We don't want to be surprised by something changing and starting to require the block being requested actually does exist in the known chain.